mesonbuild/compilers/mixins/gnu.py

```diff
@@ -91,6 +91,169 @@
     'never': ['-fdiagnostics-color=never'],
 }  # type: T.Dict[str, T.List[str]]
 
+# From common.opt in the GCC 12 source
+# Omitted non-general warnings:
+# Wabi=
+# Waggregate-return
+# Walloc-size-larger-than=BYTES
+# Walloca-larger-than=BYTES
+# Wframe-larger-than=BYTES
+# Wlarger-than=BYTES
+# Wstack-usage=BYTES
+# Wsystem-headers
+# Wtraditional
+# Wtraditional-conversion
+# Wtrampolines
+# Wvla-larger-than=BYTES
+gnu_common_warning_args = [
```

Comment:
I live `-Weverything`-esque changes. Don't like this. Seems unmaintainable. Would rather compiler developers include these with their `-Weverything` equivalent.

Reply:
Fair enough, but in GCC's case, they don't, and won't.  In reality, the problem remains.

I don't see a maintainability issue for meson - if future versions add warnings, meson _can_ add them, or it can't, and that's fine.  The point is that it's globally much more maintainable.  Doing this (via mass copy/paste) in countless projects is a nightmare - I personally have _23_ copies of this at the moment, and from a user POV, this very much smells like a thing my build system should be able to do for me.

Reply:
> Don't like this. Seems unmaintainable. Would rather compiler developers include these with their `-Weverything` equivalent.

So instead of implementing something useful now, we should ask GCC to implement it instead even though they've been steadfastly not caring for quite a while now?

On a meta level, I approve this feature using the current design for 0.64.0, hardcoding compiler flags is perfectly fine. We might discuss using something other than "4", as per discussion on IRC, but that's a separate topic.

Reply:
Multiple C compilers exist for a reason. clang has access to this option while gcc does not. Use both compilers for your CI environment.

All Meson is going to do is field issues and PRs to add warnings to these arrays. Seems like pointless noise to me.

If gcc doesn't have an equivalent, `4` should just be the equivalent of `3`. We aren't making up equivalents for compilers who don't have a `3` from what I can tell in the context of this PR. Doesn't make sense to start doing it now.

Reply:
> All Meson is going to do is field issues and PRs to add warnings to these arrays. Seems like pointless noise to me.

I made the set as objective as possible for this reason.  The only reason to add warnings here is on new major releases of GCC, which are infrequent, easy to track (new warnings are in the release notes), and don't often add warnings that aren't a part of `-Wall` or `-Wextra` anyway.  On top of that, it's not necessary for meson to update them anyway - with GCC this is _not_ like `-Weverything`, the flags are stable.

> If gcc doesn't have an equivalent, 4 should just be the equivalent of 3.

3 doesn't do what I need.

> We aren't making up equivalents for compilers who don't have a 3 from what I can tell in the context of this PR. Doesn't make sense to start doing it now.

GCC doesn't have numeric warning levels to begin with, i.e., _all_ the warning levels are already made up.

Reply:
I think it is pretty well understood in the compiler world that GCC levels of warnings go from "-Wall" to "-Wextra" to "-pedantic".

Reply:
`-Wpedantic` is more or less orthogonal to the others.

Clang and MSVC have more levels, so it already isn't true that meson warning levels exactly reflect those in compilers.  It wouldn't make sense to add this level for them, but have it do nothing for GCC because <something users don't care about>.

In any case, the whole point of this PR is to add an additional warning level.  If your argument is essentially "you don't want this", well, I promise that I do, and others do as well, and that we know what we're doing and have our reasons.

Concretely, my motivations for this are rooted in real world problems.  At the moment, I have _4767_ lines of mostly duplicated meson code that this would allow me to delete.  Meson supporting this would allow projects to use the same "all warnings then explicit suppressions" pattern with all compilers.  This seems fully in-line with the point of a build system to me: papering over differences between platforms and compilers and allowing the same patterns to work with all of them.

Reply:
I never said I don't want this. I said maintaining this huge list of flags is a burden. clang has `-Weverything` which corresponds well to `warning_level=4`. gcc doesn't. I am not convinced that maintaining these lists is Meson's burden to bare, but I am also not the person that needs to be convinced.

Reply:
I've tried to explain why I don't think it's much of a burden.  Why do you think it will be?

I understand that it's a large intimidating-looking list, and curating it was certainly a huge burden, but that's done now.  These lists only contain warnings that aren't already enabled by `-Wall -Wextra -pedantic`.  I could probably cobble together concrete statistics about that, but from skimming the release notes (and personal experience), I don't think it happens very often.

This is actually even more rare than that: meson would only "need" to update the lists if the new warning wasn't already covered by one of the existing ones.  Many of the existing flags actually enable whole groups of warnings (the set was minimized by pruning any redundancies lower in the tree, so to speak).

Anyway, worst case scenario, gcc adds some new extreme warning flags and meson doesn't update them.  Whatever.  Eager projects can add them if they want, or not.  Someone can update meson if they want, or not.  The version these come from is known and exceptions are exhaustively documented, so it's easy enough to do so.

There's no promises that meson could possibly make about this anyway.  Meson being slightly out of date relative to the current compiler is about the same as the current compiler being newer than meson.  By using `warning_level=everything`, you're explicitly signing up for all of this, so I don't think it's really a problem if meson lags or outright fails to update these lists.  Warnings at this level are not stable in any way on any compiler.

> I am not convinced that maintaining these lists is Meson's burden to bare

It's either meson, or countless other places eventually.  Meson generally encourages changes to be incorporated as special support in meson itself (often discouraging being more general in the process).  For better or worse, that's just how meson is.  It's here or nowhere.

tl;dr: It's not much of a burden, and it's not a big deal if meson fails to carry it for a while, or forever, anyway.

Reply:
I waded through the release notes back to GCC 9.  These are the only such warnings I can find:

-Wmismatched-tags
-Wredundant-tags

Both are a C++ consistency / MS portability thing, more "language stuff" if you squint right.

So, about two flags (give or take) over about 4 years, which nobody would miss.  I'd Say It's Fine™?